### ISSUE TITLE
Install and test generator-jhipster-entity-audit

### ISSUE THREAD

**Issue by pascalgrimaud** [deepu105/generator-jhipster-entity-audit#2]
I open an issue here, it will be easier than by mail because we got markdown!
I hope it's ok for you, @deepu105 

So I made some tests with our [devbox](https://github.com/jhipster/jhipster-devbox). Here all my commands. Hope it will help you...

**Step1** : clone your fork [deepu105/generator-jhipster](https://github.com/deepu105/generator-jhipster), branch module-hook
```ShellSession
mkdir -p ~/workspace/deepu
cd ~/workspace/deepu
git clone -b module-hook https://github.com/deepu105/generator-jhipster
cd generator-jhipster
npm install
npm link
ls -l /usr/lib/node_modules
```

**Step2** : clone your module [deepu105/generator-jhipster-entity-audit](https://github.com/deepu105/generator-jhipster-entity-audit)
```ShellSession
cd ~/workspace/deepu
git clone https://github.com/deepu105/generator-jhipster-entity-audit
cd generator-jhipster-entity-audit
npm install
npm link
npm link generator-jhipster
ls -l node_modules
ls -l /usr/lib/node_modules
```

**Step3** : create a sample project
```ShellSession
cd ~/workspace/deepu
mkdir 01-mysql
cd 01-mysql
npm link generator-jhipster
yo jhipster
```
I answered to all questions, all by default

**Step4** : launch your sub module
```ShellSession
$ yo jhipster-entity-audit
Composing JHipster configuration with module entity-audit
Welcome to the JHipster Entity Audit Generator! v0.1.4

events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: ENOENT: no such file or directory, scandir '.jhipster'
    at Error (native)
    at Object.fs.readdirSync (fs.js:808:18)
    at module.exports.yeoman.generators.Base.extend.initializing.getEntitityNames (/home/vagrant/workspace/deepu/generator-jhipster-entity-audit/generators/app/index.js:51:32)
    at /home/vagrant/workspace/deepu/generator-jhipster-entity-audit/node_modules/yeoman-generator/lib/base.js:430:16
    at processImmediate [as _immediateCallback] (timers.js:383:17)
```
:warning: Error, because there is no `.jhipster` directory --> maybe it's intentional

**Step5** : create simple entity
```ShellSession
$ yo jhipster:entity hello1
(...)
Everything is configured, generating the entity...

Running post run module hooks
   create .jhipster/Hello1.json
   create src/main/java/com/mycompany/myapp/domain/Hello1.java
   create src/main/java/com/mycompany/myapp/repository/Hello1Repository.java
   create src/main/java/com/mycompany/myapp/web/rest/Hello1Resource.java
   create src/main/resources/config/liquibase/changelog/20160103171718_added_entity_Hello1.xml
   create src/main/webapp/scripts/app/entities/hello1/hello1s.html
   create src/main/webapp/scripts/app/entities/hello1/hello1-detail.html
   create src/main/webapp/scripts/app/entities/hello1/hello1-dialog.html
   create src/main/webapp/scripts/app/entities/hello1/hello1-delete-dialog.html
   create src/main/webapp/scripts/app/entities/hello1/hello1.js
   create src/main/webapp/scripts/app/entities/hello1/hello1.controller.js
   create src/main/webapp/scripts/app/entities/hello1/hello1-dialog.controller.js
   create src/main/webapp/scripts/app/entities/hello1/hello1-delete-dialog.controller.js
   create src/main/webapp/scripts/app/entities/hello1/hello1-detail.controller.js
   create src/test/javascript/spec/app/entities/hello1/hello1-detail.controller.spec.js
   create src/main/webapp/scripts/components/entities/hello1/hello1.service.js
   create src/test/java/com/mycompany/myapp/web/rest/Hello1ResourceIntTest.java
   create src/main/webapp/i18n/en/hello1.json
   create src/main/webapp/i18n/fr/hello1.json
```

**Step6** : launch sub module

```ShellSession
$ yo jhipster-entity-audit
Composing JHipster configuration with module entity-audit
Welcome to the JHipster Entity Audit Generator! v0.1.4

? Do you want to enable audit for all existing entities? Yes, update all
? Do you want to add an audit log page for entities? Yes

Could not register as a post entity creation hook
Reading the JHipster project configuration for your module

Updating selected entities Hello1

Make sure these classes does not extend any other class to avoid any errors during compilation.
errorError: ENOENT: no such file or directory, open '/home/vagrant/workspace/deepu/generator-jhipster-entity-audit/generators/app/templates/src/main/webapp/i18n/en/_Audits.json'
   create src/main/java/com/mycompany/myapp/config/audit/AsyncEntityAuditEventWriter.java
   create src/main/java/com/mycompany/myapp/config/audit/EntityAuditEventListener.java
   create src/main/java/com/mycompany/myapp/config/audit/EntityAuditAction.java
   (...)
```

:warning: There is one error here. See logs.
And the `.jhipster-modules.json` file is empty too.

**Step 7** : I modify manually the .jhipster-modules.json
```yaml
[
  {
    "name": "Entity Audit",
    "npmPackageName": "generator-jhipster-entity-audit",
    "description": "Add support for entity audit and audit log page",
    "hookFor" : "entity",
    "hookType" : "post",
    "generatorCallback" : "jhipster-entity-audit:entity"
  }
]
```

**Step 8** : I create a new entity, it should be hook by your sub module
```ShellSession
$ yo jhipster:entity world1
The entity world1 is being created.
Generating field #1
? Do you want to add a field to your entity? Yes
? What is the name of your field? name
? What is the type of your field? String
? Do you want to add validation rules to your field? No
=================World1=================
name (String)
Generating field #2
? Do you want to add a field to your entity? No
=================World1=================
name (String)
Generating relationships with other entities
? Do you want to add a relationship to another entity? No
===========World1==============
name (String)
-------------------
? Do you want to use a Data Transfer Object (DTO)? No, use the entity directly
? Do you want to use separate service class for your business logic? No, the REST controller should use the repository directly
? Do you want pagination on your entity? No
Everything is configured, generating the entity...

Running post run module hooks
Composing JHipster configuration with module entity-audit
Running JHipster Entity Audit Generator! v0.1.4

? Do you want to enable audit for this entity? Yes
Reading the JHipster project configuration for your module

Updating the entity for audit World1

Unable to find src/main/java/com/mycompany/myapp/domain/World1.java or missing required pattern. File rewrite failed.
Error: ENOENT: no such file or directory, open '/home/vagrant/workspace/deepu/01-mysql/src/main/java/com/mycompany/myapp/domain/World1.java'

Unable to find undefined or missing required jhipster-needle. Column not added.
TypeError: Path must be a string. Received undefined
   create .jhipster/World1.json
   create src/main/java/com/mycompany/myapp/domain/World1.java
   create src/main/java/com/mycompany/myapp/repository/World1Repository.java
   create src/main/java/com/mycompany/myapp/web/rest/World1Resource.java
   create src/main/resources/config/liquibase/changelog/20160103172323_added_entity_World1.xml
   create src/main/webapp/scripts/app/entities/world1/world1s.html
   create src/main/webapp/scripts/app/entities/world1/world1-detail.html
   create src/main/webapp/scripts/app/entities/world1/world1-dialog.html
   create src/main/webapp/scripts/app/entities/world1/world1-delete-dialog.html
   create src/main/webapp/scripts/app/entities/world1/world1.js
   create src/main/webapp/scripts/app/entities/world1/world1.controller.js
   create src/main/webapp/scripts/app/entities/world1/world1-dialog.controller.js
   create src/main/webapp/scripts/app/entities/world1/world1-delete-dialog.controller.js
   create src/main/webapp/scripts/app/entities/world1/world1-detail.controller.js
   create src/test/javascript/spec/app/entities/world1/world1-detail.controller.spec.js
   create src/main/webapp/scripts/components/entities/world1/world1.service.js
   create src/test/java/com/mycompany/myapp/web/rest/World1ResourceIntTest.java
   create src/main/webapp/i18n/en/world1.json
   create src/main/webapp/i18n/fr/world1.json

End of entity-audit
```

:warning: again, I got error.

**Comment by deepu105** [comment]
Ok these are new errors which im not seeing when i run the same steps in my
laptop, the only difference here seems to be the OS, let me check those
again

**Comment by pascalgrimaud** [comment]
We can have same OS => [devbox](https://github.com/jhipster/jhipster-devbox)

Have a look this morning at your code.
You tried to introduce a very nice mecanism :+1:

What I see is, when launching for example : `yo jhipster:entity bonjour` (on an existing entity) :
- after answering all questions
- your sub module is launch first
- then, the main entity generator is launch, so it erases all your modif
=> So I finally manage to reproduce your issue.

I think there is something to play with `this.async()`
I will go on to analyze as soon as I have time...

**Comment by deepu105** [comment]
yes i think so something to do with run loops more than the async i guess im gonna try something on that line

**Comment by deepu105** [comment]
I have corrected all the errors you mentioned above but the issue still persists I have opened a ticket in yeoman project as well https://github.com/yeoman/yeoman/issues/1582

**Comment by deepu105** [comment]
@pascalgrimaud @cbornet @gmarziou @jdubois I have debugged this for a while and I think here is what happens.

Yeoman puts all the methods in a run queue, the methods are sorted on priority and priority is by method name, so they have this names like install, writing, promting etc which will be run in certain priority and any methods with this name will be run in that particular order, methods can be lumbed together inside a priorityName to ensure order, every other method will be running in default order.

Our entity generator doesn't follow this naming and all methods are run in default order which i doubt is after writing method and hence, the modules writing is done before the entity generator file method.

Im gonna try to refactor the entity gen index.js to adher to the naming convention to see if the issue gets resolved.

**Comment by pascalgrimaud** [comment]
My bad... :weary: I didn't notice our entity generator doesn't follow this naming.
I read this http://yeoman.io/authoring/running-context.html when I began to write my first module and I used this order : initializing -> prompting -> writing -> install -> end

**Comment by deepu105** [comment]
yes and if im not wrong default methods are run between prompting -> writing but not very sure, it looks like so from debug logs

**Comment by deepu105** [comment]
I have rewrote entity gen to follow this order will put in a PR later

**Comment by deepu105** [comment]
sorry this is correct order

initializing - Your initialization methods (checking current project state, getting configs, etc)
prompting - Where you prompt users for options (where you'd call this.prompt())
configuring - Saving configurations and configure the project (creating .editorconfig files and other metadata files)
`default` - If the method name doesn't match a priority, it will be pushed to this group.
writing - Where you write the generator specific files (routes, controllers, etc)
conflicts - Where conflicts are handled (used internally)
install - Where installation are run (npm, bower)
end - Called last, cleanup, say good bye, etc

**Comment by deepu105** [comment]
And voila that was the issue its working now :joy:

**Comment by deepu105** [comment]
```
initializing - Your initialization methods (checking current project state, getting configs, etc)
prompting - Where you prompt users for options (where you'd call this.prompt())
configuring - Saving configurations and configure the project (creating .editorconfig files and other metadata files)
default - If the method name doesn't match a priority, it will be pushed to this group.
writing - Where you write the generator specific files (routes, controllers, etc)
conflicts - Where conflicts are handled (used internally)
install - Where installation are run (npm, bower)
end - Called last, cleanup, say good bye, etc
```

This is the new order

**Comment by deepu105** [comment]
This is the order in which entity generator runs now
```
yeoman:generator Queueing getConfig in initializing +0ms
  yeoman:generator Queueing validateEntityName in initializing +2ms
  yeoman:generator Queueing setupVars in initializing +0ms
  yeoman:generator Queueing askForFields in prompting +0ms
  yeoman:generator Queueing askForRelationships in prompting +0ms
  yeoman:generator Queueing askForDTO in prompting +0ms
  yeoman:generator Queueing askForService in prompting +0ms
  yeoman:generator Queueing askForPagination in prompting +0ms
  yeoman:generator Queueing setup in writing +0ms
  yeoman:generator Queueing validate in writing +0ms
  yeoman:generator Queueing LoadInMemoryData in writing +0ms
  yeoman:generator Queueing insight in writing +0ms
  yeoman:generator Queueing writeEnumFiles in writing +0ms
  yeoman:generator Queueing writeServerFiles in writing +0ms
  yeoman:generator Queueing writeDbFiles in writing +0ms
  yeoman:generator Queueing writeClientFiles in writing +0ms
  yeoman:generator Queueing writeTestFiles in writing +0ms
  yeoman:generator Queueing default in default +0ms
  yeoman:generator Queueing afterRunHook in end +0ms
  yeoman:generator Running getConfig +16ms
Found the .jhipster/Label.json configuration file, automatically generating the entity
  yeoman:generator Running validateEntityName +0ms
  yeoman:generator Running setupVars +0ms
The entity label is being created.
  yeoman:generator Running askForFields +16ms
  yeoman:generator Running askForRelationships +0ms
  yeoman:generator Running askForDTO +0ms
  yeoman:generator Running askForService +0ms
  yeoman:generator Running askForPagination +0ms
  yeoman:generator Running default +0ms
  yeoman:generator Running setup +0ms
  yeoman:generator Running validate +0ms
  yeoman:generator Running LoadInMemoryData +15ms
  yeoman:generator Running insight +0ms
  yeoman:generator Running writeEnumFiles +0ms
  yeoman:generator Running writeServerFiles +0ms
  yeoman:generator Running writeDbFiles +54ms
  yeoman:generator Running writeClientFiles +15ms
  yeoman:generator Running writeTestFiles +85ms
 conflict src\main\java\com\mycompany\myapp\domain\Label.java
? Overwrite src\main\java\com\mycompany\myapp\domain\Label.java? (Ynaxdh)
? Overwrite src\main\java\com\mycompany\myapp\domain\Label.java? overwrite
    force src\main\java\com\mycompany\myapp\domain\Label.java
identical src\main\java\com\mycompany\myapp\repository\LabelRepository.java
identical src\main\java\com\mycompany\myapp\repository\search\LabelSearchRepository.java
identical src\main\java\com\mycompany\myapp\web\rest\LabelResource.java
identical src\main\java\com\mycompany\myapp\service\LabelService.java
 conflict src\main\resources\config\liquibase\changelog\20150805124936_added_entity_Label.xml
? Overwrite src\main\resources\config\liquibase\changelog\20150805124936_added_entity_Label.xml? overwrite
    force src\main\resources\config\liquibase\changelog\20150805124936_added_entity_Label.xml
identical src\main\webapp\scripts\app\entities\label\labels.html
identical src\main\webapp\scripts\app\entities\label\label-detail.html
identical src\main\webapp\scripts\app\entities\label\label-dialog.html
identical src\main\webapp\scripts\app\entities\label\label-delete-dialog.html
identical src\main\webapp\scripts\app\entities\label\label.js
identical src\main\webapp\scripts\app\entities\label\label.controller.js
identical src\main\webapp\scripts\app\entities\label\label-dialog.controller.js
identical src\main\webapp\scripts\app\entities\label\label-delete-dialog.controller.js
identical src\main\webapp\scripts\app\entities\label\label-detail.controller.js
identical src\test\javascript\spec\app\entities\label\label-detail.controller.spec.js
identical src\main\webapp\scripts\components\entities\label\label.service.js
identical src\main\webapp\scripts\components\entities\label\label.search.service.js
identical src\main\webapp\i18n\en\label.json
identical src\main\webapp\i18n\fr\label.json
identical src\test\java\com\mycompany\myapp\web\rest\LabelResourceIntTest.java
identical src\test\gatling\simulations\LabelGatlingTest.scala
  yeoman:generator Running afterRunHook +3m

Running post run module hooks
  yeoman:generator Queueing compose in initializing +0ms
  yeoman:generator Queueing displayLogo in initializing +2ms
  yeoman:generator Queueing prompting in prompting +2ms
  yeoman:generator Queueing writing in writing +3ms
  yeoman:generator Queueing end in end +1ms
  yeoman:generator Running compose +3ms
Composing JHipster configuration with module entity-audit
  yeoman:generator Queueing configurer in default +1s
  yeoman:generator Running displayLogo +0ms
Running JHipster Entity Audit Generator! v0.1.9

  yeoman:generator Running prompting +16ms
? Do you want to enable audit for this entity? (Y/n) y
? Do you want to enable audit for this entity? Yes
  yeoman:generator Running configurer +8s
Reading the JHipster project configuration for your module
  yeoman:generator Running writing +8s

I'm updating the entity for audit Label
  yeoman:generator Running end +35ms

End of entity-audit
  yeoman:generator Running the hooked generators +38ms
  yeoman:generator Running the hooked generators +3ms
  yeoman:generator Running the hooked generators +0ms

```

**Comment by pascalgrimaud** [comment]
nice :+1:

**Comment by cbornet** [comment]
:clap:

**Comment by deepu105** [comment]
ill commit these soon :+1:

**Comment by deepu105** [comment]
https://github.com/jhipster/generator-jhipster/pull/2569

**Comment by jdubois** [comment]
Yes, great job!
This has totally changed since the first release of JHipster, and this is
why we don't follow this at all.

On Tue, Jan 5, 2016 at 9:32 AM, Deepu k Sasidharan <notifications@github.com
> wrote:

> ill commit these soon [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/deepu105/generator-jhipster-entity-audit/issues/2#issuecomment-168934059>
> .
>



-- 
Julien Dubois

Twitter: @juliendubois <http://twitter.com/#!/juliendubois>
Phone: +33 (0)6 25 02 34 18

**Comment by deepu105** [comment]
@jdubois changing the structure has resolved the issue I had with conflicts in file writing, now my audit module can be used as post entity hook :) @cbornet I guess your module will work as well, Ill publish a release for the module now

**Comment by deepu105** [comment]
version 1.0 released, but need to wait for next JHipster release :(

**Comment by deepu105** [comment]
closed via jhipster/generator-jhipster#2569
